src/coreclr/jit/flowgraph.cpp

```diff
@@ -2803,6 +2803,62 @@ bool Compiler::fgSimpleLowerCastOfSmpOp(LIR::Range& range, GenTreeCast* cast)
     return false;
 }
 
+//------------------------------------------------------------------------
+// fgSimpleLowerBswap16 : Optimization to remove CAST nodes from operands of small ops that depents on
+// lower bits only (currently only BSWAP16).
+// Example:
+//      BSWAP16(CAST(x)) transforms to BSWAP16(x)
+//
+// Returns:
+//      True or false, representing changes were made.
+//
+// Notes:
+//      This optimization could be done in morph, but it cannot because there are correctness
+//      problems with NOLs (normalized-on-load locals) and how they are handled in VN.
+//      Simple put, you cannot remove a CAST from CAST(LCL_VAR{nol}) in HIR.
+//
+//      Because the optimization happens during rationalization, turning into LIR, it is safe to remove the CAST.
+//
+bool Compiler::fgSimpleLowerBswap16(LIR::Range& range, GenTree* op)
+{
+    assert(op->OperIs(GT_BSWAP16));
+
+    if (opts.OptimizationDisabled())
+        return false;
+
+    // When openrand is a integral cast
+    // When both source and target sizes are at least the operation size
+    var_types opSize = TYP_SHORT;
+
+    bool madeChanges = false;
+
+    if (op->gtGetOp1()->OperIs(GT_CAST))
+    {
+        GenTreeCast* op1          = op->gtGetOp1()->AsCast();
+        var_types    castFromType = op1->CastFromType();
+        var_types    castToType   = op1->CastToType();
+
+        if (!op1->gtOverflow() && varTypeIsIntegralOrI(castFromType) &&
+            genTypeSize(castFromType) >= genTypeSize(opSize) && genTypeSize(castToType) >= genTypeSize(opSize))
+        {
+            // Remove the cast.
```

Comment:
```suggestion
        if (!op1->gtOverflow() && (genActualType(op1->CastFromType()) == TYP_INT) && (genTypeSize(op1->CastToType()) >= 2))
        {
            // This cast does not affect the lower 16 bits. It can be removed.
```

Seems a bit clearer to me (also, the `castFromType` check is not correct as is -- it would create discard the cast even for type changing casts like `long -> int`).

Reply:
I tested with `Reverse(ulong value) => BinaryPrimitives.ReverseEndianness((ushort)value)`, and it can cancel the `long -> short` cast. By illegal IR, do you mean it's not the correct stage to cancel the cast?

Reply:
`BSWAP16` as a node does not expect to see `TYP_LONG` operands, only `TYP_INT` operands. If it happens to work it is by chance. To optimize a cast like that, the cast should be changed to `long -> int` instead of removed.

Reply:
It's indeed a deoptimization to change the cast to `int`, because the zero extension after BSWAP16 can't be eliminated.

Reply:
I've updated this with only the simple `int -> short` case preserved.